### PR TITLE
lib/alloc_prof.[ch]: add simple memory allocator profiler.

### PR DESCRIPTION
**Pull request by nikitadanilov**
See lib/alloc_prof.h for the documentation.

Signed-off-by: Nikita Danilov <nikita.danilov@seagate.com>

-----
[View rendered doc/fault-injection.md](https://github.com/Seagate/cortx-motr/blob/alloc-profile/doc/fault-injection.md)

### CODE REVIEW

**Comment by nikitadanilov** [comment]
As an example, the allocator profile output for the full UT run (usesock branch):
```
$ sort -k3 -n ~/tmp/usesock-ut-alloc-prof | tail
     alloc      19833  403433688   120 cas/service.c 671 cas_fom_create repv
     alloc       7047  451008000   859 net/sock/sock.c 4696 get_fd fd->md_buf.u_sock
     alloc      32969  815257432    61 ha/msg_queue.c 58 m0_ha_msg_queue_alloc qitem
     alloc     256795  922017611  5382 lib/memory.c 128 m0_alloc m0_alloc
     alloc      20649 2079782152    35 be/io.c 192 m0_be_io_allocate bv->ov_vec.v_count
     alloc      20649 2079782152    35 be/io.c 193 m0_be_io_allocate bv->ov_buf
     alloc      20649 2079782152    35 be/io.c 194 m0_be_io_allocate iv->iv_vec.v_count
     alloc      20649 2079782152    35 be/io.c 195 m0_be_io_allocate iv->iv_index
     alloc       5287 4160343280     3 be/fmt.c 259 m0_be_fmt_group_init cheader->fch_reg_area.chr_reg
     alloc       5287 4160343280     3 be/fmt.c 267 m0_be_fmt_group_init content->fmc_reg_area.cra_reg

**Comment by andriytk** [comment]
@nikitadanilov, https://cbea.ms/git-commit/#end

**Comment by andriytk** [comment]
> As an example, the allocator profile output for the full UT run (usesock branch):

@nikitadanilov, it would be useful to add a few comments on the result of this profiling - what the numbers mean, how to interpret them, etc.

**Review by andriytk — APPROVED** [review]

File: lib/alloc_prof.c [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+enum { SLOTS = 1000 };
+
+extern char __start_m0allocprof;
+extern char __stop_m0allocprof;
+
+static struct m0_alloc_prof *begin = (void *)&__start_m0allocprof;
+static struct m0_alloc_prof *end = (void *)&__stop_m0allocprof;
+
+static struct m0_alloc_prof_tls ptdata[AP_NR][SLOTS] = {};
```
Comment (shashank-parulekar) [inline lib/alloc_prof.c:45]: This variable is supposed to be per thread according to the documentation for struct m0_alloc_prof_tls.

File: lib/memory.h [hunk 2]
```diff
@@ -175,6 +180,11 @@ M0_INTERNAL bool m0_is_poisoned(const void *p);
  */
 M0_INTERNAL int m0_dont_dump(void *p, size_t size);
 
+/**
+ * Allocator entry point used within M0_ALLOC_{PTR,ARR}() for already profiled
+ * allocations.
+ */
+void *m0_alloc_profiled(size_t size, struct m0_alloc_prof *ap);
```
Comment (shashank-parulekar) [inline lib/memory.h:187]: Compile error is seen for this line on current code base.

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: lib/alloc_prof.c [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+static struct m0_mutex guard;
+
+/** Updates the local counters. */
+void m0_alloc_prof_update(int atype, struct m0_alloc_prof *ap, size_t nob)
+{
+	struct m0_alloc_prof_tls *pt = &ptdata[atype][ap - begin];
```
[the remark this answers is not in the document]
Comment (nikitadanilov) [reply lib/alloc_prof.c:51]: It seems you are looking at the alloc-profile branch and not at the PR. I made some further experimental undebugged changes to the branch. Please review the commits in the PR only.

**Review by huanghua78 — APPROVED** [review]

**Pull request closed without merging**